Undeploy models with no WorkerNodes

brianf-aws opened the pull request:
This PR  aims to undeploy modelIds that have no nodes associated to them so as to keep the intention of undeploy truthful.

### Description
When performing undeploy if the model has no nodes associated to it then it will reset the index to UNDEPLOY status

Here is an example of why this code change is needed

This secnario is for the `PARTIALLY_DEPLOYED` issue.
1. Have nodes a,b,c,d in cluster associated with `modelID:@` i.e. peform deploy on it
2. Bring a,b down while having the syncup job running
3. By Now sync up will make this PARTIALLY_UNDEPLOYED
4. stop sync up
5. Bring other 2 c,d down and bring 2 nodes (these now have new ids 1,2)
6. bring the other two nodes back which have different ids so now the cluster has (1,2,3,4)
But the model index says `PARTIALLY_DEPLOYED` and no nodes are servicing

This code fix says. If no nodes are servicing this model then I need to set the index to UNDEPLOYED no matter if its already UNDEPLOYED or not.

### Related Issues
Resolves https://github.com/opensearch-project/ml-commons/issues/3285
<!-- List any other related issues here -->

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -157,10 +163,36 @@ private void undeployModels(String[] targetNodeIds, String[] modelIds, ActionLis
+        bulkUpdateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        log.info("No models service: {}", modelIds.toString());
+        client.bulk(bulkUpdateRequest, ActionListener.wrap(br -> { log.debug("Successfully set modelIds to UNDEPLOY in index"); }, e -> {
+            log.error("Failed to set modelIds to UNDEPLOY in index", e);
```
Should we send this exception back to client side ? If yes, we should pass listener to this method and add this line here
```
listener.onFailure(e);
```

dhrubo-os replied:
+1

brianf-aws replied:
I'm not sure the user is concerned with the failure? The only way I see this as a problem if the model index does not exist and user does undeploy, this might cause a write issue.

mingshl commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -157,10 +163,36 @@ private void undeployModels(String[] targetNodeIds, String[] modelIds, ActionLis
         MLUndeployModelNodesRequest mlUndeployModelNodesRequest = new MLUndeployModelNodesRequest(targetNodeIds, modelIds);
 
         client.execute(MLUndeployModelAction.INSTANCE, mlUndeployModelNodesRequest, ActionListener.wrap(r -> {
+            if (r.getNodes().isEmpty()) {
```
it make sense that when execute undeploy model, the response return no worder nodes, then set the model to undeploy. 

but this doesn't fix the partially undeployed issue, right?

brianf-aws replied:
`PARTIALLY_UNDEPLOYED` is a bit of a mixture of different scenarios one of them like so 
1. Have nodes a,b,c,d in cluster associated with `modelID:@` i.e. peform deploy on it
2. Bring a,b down while having the syncup job running
3. By Now sync up will make this PARTIALLY_UNDEPLOYED
4. stop sync up
5. Bring other 2 c,d down and bring 2 nodes (these now have new ids 1,2)
6. bring the other two nodes back which have different ids so now the cluster has (1,2,3,4)
But the model index says `PARTIALLY_DEPLOYED` and no nodes are servicing

This code fix says. If no nodes are servicing this model then I need to set the index to UNDEPLOYED no matter if its already UNDEPLOYED or not.

dhrubo-os commented:
Can we add unit test?

Zhangxunmt commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -157,10 +163,36 @@ private void undeployModels(String[] targetNodeIds, String[] modelIds, ActionLis
+        }
+        bulkUpdateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        log.info("No models service: {}", modelIds.toString());
+        client.bulk(bulkUpdateRequest, ActionListener.wrap(br -> { log.debug("Successfully set modelIds to UNDEPLOY in index"); }, e -> {
```
I think you can return MLUndeployModelsResponse including the original nodes rather than empty. You can find some examples in the tests how to create a  new MLUndeployModelsResponse.

ylwu-amzn replied:
If we don't return in `br -> { log.debug("Successfully set modelIds to UNDEPLOY in index"); }`, it's possible that when client side receive the undeploy response, the model still on DEPLOYED state.

brianf-aws replied:
Im not sure how I feel about creating a new one based on the failures,I think it will be misleading.  I will pass the original response r instead.

Zhangxunmt replied:
The failures should just return a message. But for the success case we should return something rather than {}

dhrubo-os replied:
if success case shows:

```
{
"node id 1" : Not Found,
"node id 2" : Not Found
}
```

I don't think this will add much value from customer's POV.

rbhavna replied:
This will still give empty response which is not accurate. Since we cannot send nodes as response in this case, lets send something to show model/models undeployed successfully. Something like

```
{
<model_id_1>: "UNDEPLOYED SUCCESSFULLY",
<model_id_2>: "UNDEPLOYED SUCCESSFULLY"
}
```

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -157,10 +163,36 @@ private void undeployModels(String[] targetNodeIds, String[] modelIds, ActionLis
 
         client.execute(MLUndeployModelAction.INSTANCE, mlUndeployModelNodesRequest, ActionListener.wrap(r -> {
+            if (r.getNodes().isEmpty()) {
+                bulkSetModelIndexToUndeploy(modelIds);
```
```suggestion
                bulkSetModelIndexToUndeploy(modelIds, listener);
                return;
```

brianf-aws replied:
Thanks added the listener and response and returned early

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -156,11 +163,48 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLUnde
+        bulkUpdateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        log.info("No nodes service: {}", modelIds.toString());
+
+        client.bulk(bulkUpdateRequest, ActionListener.wrap(br -> {
```
Wrap this with 
```
try (ThreadContext.StoredContext threadContext = client.threadPool().getThreadContext().stashContext()) {
```
Refer to https://github.com/opensearch-project/ml-commons/blob/main/memory/src/main/java/org/opensearch/ml/memory/index/ConversationMetaIndex.java#L161

brianf-aws replied:
Thanks for the insight, Just added this change

brianf-aws commented:
> Can we add unit test?

Added 2 UTs for code fix
1. Check that the bulk write occurred when undeploy returned {}. This is a sign that the stale model index is `UNDEPLOYED`
2. Check that bulk write did not occur when some nodes have a response to the model, undeploy occured and changed index.

https://github.com/opensearch-project/ml-commons/blob/22b558d1f985d6de2c7b4713a6928fde73ea0030/plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelAction.java#L192-L195

dhrubo-os commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -156,11 +163,48 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLUnde
-        client.execute(MLUndeployModelAction.INSTANCE, mlUndeployModelNodesRequest, ActionListener.wrap(r -> {
-            listener.onResponse(new MLUndeployModelsResponse(r));
+        client.execute(MLUndeployModelAction.INSTANCE, mlUndeployModelNodesRequest, ActionListener.wrap(response -> {
+            if (response.getNodes().isEmpty()) {
```
[nit]Let's add a detailed explanation here to ensure that future developers understand the reason behind this logic and aren't confused about why it's implemented this way.

ylwu-amzn commented on plugin/src/main/java/org/opensearch/ml/action/undeploy/TransportUndeployModelsAction.java:
```diff
@@ -156,11 +164,68 @@ protected void doExecute(Task task, ActionRequest request, ActionListener<MLUnde
+        }
+
+        bulkUpdateRequest.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE);
+        log.info("No nodes service: {}", Arrays.toString(modelIds));
```
How about clarifying more ?
```suggestion
        log.info("No nodes running these models: {}", Arrays.toString(modelIds));
```

brianf-aws replied:
Valid feedback added to commit [77f6e5b](https://github.com/opensearch-project/ml-commons/pull/3380/commits/77f6e5bbb837191fe804ee26c5abd89a9942727e)

mingshl commented:
@brianf-aws you need to rebase first

mingshl commented:
LGTM. I added 2.x backport label, did you figure out what other versions you need to backport to? I will add the labels for you

brianf-aws commented:
> LGTM. I added 2.x backport label, did you figure out what other versions you need to backport to? I will add the labels for you

Thank you! Yes I think we can do OS.2.11 - 2.18

brianf-aws commented:
Sorry can we make this to 2.x ? since this works of multi-tenancy and would break every other backport

brianf-aws commented:
Looks like an IT is causing an issue
```
RestBedRockInferenceIT > test_bedrock_multimodal_model_empty_imageInput_null_textInput STANDARD_OUT
    [2025-01-28T13:06:27,997][INFO ][o.o.m.r.RestBedRockInferenceIT] [test_bedrock_multimodal_model_empty_imageInput_null_textInput] before test
Error: 25-01-28T13:06:48,987][ERROR][o.o.m.r.MLCommonsRestTestCase] [test_bedrock_multimodal_model_empty_imageInput_null_textInput] method [POST], host [http://[::1]:38167], URI [/_plugins/_ml/_predict/TEXT_EMBEDDING/UFOGrpQB2p_oggmIDsWU], status line [HTTP/1.1 400 Bad Request]
    {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"No input text or image provided"}],"type":"illegal_argument_exception","reason":"No input text or image provided"},"status":400}
    org.opensearch.client.ResponseException: method [POST], host [http://[::1]:38167], URI [/_plugins/_ml/_predict/TEXT_EMBEDDING/UFOGrpQB2p_oggmIDsWU], status line [HTTP/1.1 400 Bad Request]
    {"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"No input text or image provided"}],"type":"illegal_argument_exception","reason":"No input text or image provided"},"status":400}
    	at org.opensearch.client.RestClient.convertResponse(RestClient.java:501) ~[opensearch-rest-client-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
    	at org.opensearch.client.RestClient.performRequest(RestClient.java:384) ~[opensearch-rest-client-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
    	at org.opensearch.client.RestClient.performRequest(RestClient.java:359) ~[opensearch-rest-client-3.0.0-SNAPSHOT.jar:3.0.0-SNAPSHOT]
    	at org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:183) ~[test/:?]
    	at org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:156) ~[test/:?]
    	at org.opensearch.ml.utils.TestHelper.makeRequest(TestHelper.java:145) ~[test/:?]
    	at org.opensearch.ml.rest.MLCommonsRestTestCase.predictTextEmbeddingModel(MLCommonsRestTestCase.java:925) ~[test/:?]
    	at org.opensearch.ml.rest.RestBedRockInferenceIT.test_bedrock_multimodal_model_empty_imageInput_null_textInput(RestBedRockInferenceIT.java:233) ~[test/:?]
```

Is this a flaky known issue if not can we merge? cc: @jngz-es, @mingshl
